akka-stream-tests/src/test/scala/akka/stream/scaladsl/AttributesSpec.scala

```diff
@@ -3,70 +3,464 @@
+  "attributes on a source" must {
 
-    "give access to first attribute" in {
-      attributes.getFirst[Name] should ===(Some(Attributes.Name("a")))
+    "make the attributes on fromGraph(single-source-stage) Source behave the same as the stage itself" in {
+      val attributes =
+        Source.fromGraph(new AttributesSource(Attributes.name("original-name") and whateverAttribute("whatever")))
+          .withAttributes(Attributes.name("replaced")) // replaces all
+          .toMat(Sink.head)(Keep.left).withAttributes(Attributes.name("whole-graph"))
+          .run()
+
+      // most specific
+      attributes.mostSpecific[Name] should contain(Name("replaced"))
+      attributes.mostSpecific[WhateverAttribute] shouldBe empty
+
+      // least specific
+      attributes.leastSpecific[Name] should contain(Name("whole-graph"))
+      attributes.leastSpecific[WhateverAttribute] shouldBe empty
     }
 
-    "give access to attribute byt type" in {
-      attributes.get[Name] should ===(Some(Attributes.Name("b")))
+    "not replace stage specific attributes with attributes on surrounding composite source" in {
+      val attributes = Source.fromGraph(new AttributesSource(Attributes.name("original-name")))
+        .map(identity)
+        .addAttributes(Attributes.name("composite-graph"))
+        .toMat(Sink.head)(Keep.left)
+        .run()
+
+      // most specific still the original as the attribute was added on the composite source
+      attributes.mostSpecific[Name] should contain(Name("original-name"))
+
+      // least specific
+      attributes.leastSpecific[Name] should contain(Name("composite-graph"))
+    }
+
+    "use the initial attribute as default to select dispatcher" in {
+      val dispatcher =
+        Source.fromGraph(new ThreadNameSnitchingStage("my-dispatcher"))
+          .runWith(Sink.head)
+          .futureValue
+
+      dispatcher should startWith("AttributesSpec-my-dispatcher")
+    }
+
+    "use an explicit attribute on the stage to select dispatcher" in {
+      val dispatcher =
+        Source.fromGraph(
+          // directly on stage
+          new ThreadNameSnitchingStage("akka.stream.default-blocking-io-dispatcher")
+            .addAttributes(ActorAttributes.dispatcher("my-dispatcher")))
+          .runWith(Sink.head)
+          .futureValue
+
+      dispatcher should startWith("AttributesSpec-my-dispatcher")
+    }
+
+    "use the most specific dispatcher when another one is defined on a surrounding composed graph" in {
+      val dispatcher =
+        Source.fromGraph(
+          new ThreadNameSnitchingStage("akka.stream.default-blocking-io-dispatcher"))
+          .map(identity)
+          // this is now for the composed source -> flow graph
+          .addAttributes(ActorAttributes.dispatcher("my-dispatcher"))
+          .runWith(Sink.head)
+          .futureValue
+
+      dispatcher should startWith("AttributesSpec-akka.stream.default-blocking-io-dispatcher")
```

Comment:
add a test for this scenario but the my-dispatcher being `.important` I assume it works fine right? It's also solving the use-cases I had in mind then

Reply:
D'oh, I added this test which fails, and I think this is the one I was researching a long time ago, where the materializer has already introduced an async boundary, because it saw a differing dispatcher. So this is another bug that needs to be fixed.